File: example/external/c/yarpread.c
/*
 * SPDX-FileCopyrightText: 2006-2021 Istituto Italiano di Tecnologia (IIT)
 * SPDX-FileCopyrightText: 2006-2010 RobotCub Consortium
 * SPDX-License-Identifier: BSD-3-Clause
 */

#include "yarpmin.h"

#include <stdio.h>
#include <stdlib.h>

int main(int argc, char *argv[]) {
    char buf[1000];
    char *pending_buf = NULL;
    int pending_len = 0;
    yarpAddress addr;
    yarpConnection con;
    int res;
    if (argc<2) {
        printf("Call as:\n  %s /port/to/read/from\n",
               argv[0]);
        exit(1);
    }
    if (argc==4) {
        yarp_init_with(argv[2],atoi(argv[3]));
    } else {
        yarp_init();
    }
    res = yarp_port_lookup(&addr,argv[1]);
    if (res<0) {
        printf("Failed to find port\n");
        exit(1);
    }
    fprintf(stderr,"Connecting to %s:%d\n", addr.host, addr.port_number);
    con = yarp_prepare_to_read(&addr);
    if (!yarp_is_valid(con)) {
        printf("Connection failed\n");
        exit(1);
    }
    while (res>=0) {
        res = yarp_receive_lines(con,buf,sizeof(buf),&pending_buf,&pending_len);
        if (res<=0) break;
        printf("%s\n", buf);
    }
    yarp_disconnect(con);
    yarp_fini();
    return 0;
}

File: example/external/python/yarpread.py
#!/usr/bin/python

# Copyright (C) 2006-2021 Istituto Italiano di Tecnologia (IIT)
# Copyright (C) 2006-2010 RobotCub Consortium
# All rights reserved.
#
# This software may be modified and distributed under the terms of the
# BSD-3-Clause license. See the accompanying LICENSE file for details.

import socket, re, sys

if len(sys.argv)!=2:
    print 'Call as:\n  %s /port/to/read/from'%sys.argv[0]
    exit(1)

try:
    import find_name_server
    name_server = find_name_server.find_name_server()
    print "Nameserver is here:", name_server
except:
    name_server = ('localhost',10000)
    print "Assuming nameserver is here:", name_server

def get_text(sock): # get a single line of text from a socket
    result = ""
    while result.find('\n')==-1:
        result = result + sock.recv(1024)
    result = re.sub('[\r\n].*','',result)
    return result
def get_address(s): # extract address from a name server response
    m = re.match("registration name [^ ]+ ip ([^ ]+) port ([0-9]+) type tcp",s)
    return (m.group(1),int(m.group(2))) if m else None
def send_message(addr,message): # send a message and expect a reply
    sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
    sock.connect(addr)
    sock.send('CONNACK extern\n')
    get_text(sock)
    sock.send('d\n%s\n' % message)
    result = get_text(sock)
    sock.close()
    return result

def read_loop(addr):
    sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
    sock.connect(addr)
    sock.send('CONNECT extern\nr\n')
    while True:
        data = sock.recv(1024)
        if not data: break
        print data,

port_name = sys.argv[1]
read_loop(get_address(send_message(name_server,"query %s"%port_name)))
